eip/EIPS/eip-4337.md

```diff
@@ -271,23 +280,27 @@ In order to distinguish between them, there is a call to the NUMBER opcode (`blo
-6. Any `CALL` or `CALLCODE` during validation has `value=0`, except for the transfer from the wallet to the entry point.
-7. No `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` results in an out-of-gas revert.
-8. No `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` to addresses with `EXTCODESIZE=0`.
-9. Any `GAS` opcode is followed immediately by one of { `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` }.
-10. `EXTCODEHASH` of every address accessed (by any opcode) does not change between first and second simulations of the op.
-11. If `op.initcode.length != 0` , allow only one `CREATE2` opcode call (in the validateUserOp block), otherwise forbid `CREATE2`.
-
-Since the wallet is allowed to access its own entry point deposit in order to top it up when needed, the client must know the storage slot in order to whitelist it. The entry point therefore implements the following view function:
-
-```c++
-function getSenderStorage(address sender) external view returns (uint256[] memory senderStorageCells)
-```
+2. The first (validateUserOp) call is allowed to access storage slot of the wallet itself (see Storage access by Slots, below), 
+   but only on contracts that are not part of the current bundle (that is, not in a paymaster or any other wallet)
+   Note that the wallet is allowed (and expected) to deposit to its own balance in the EntryPoint. That is allowed by this rule.
+3. The second (validatePaymasterUserOp), the paymaster may access storage slots of itself AND of the wallet, above.
+   A bundler MAY whitelist specific paymaster addresses, and not enforce this storage access limitation.
+4. The third (validateUserOpSignature) view-call, the aggregator is allowed to access storage slots of itself and of the wallet. 
+5. Any `CALL` or `CALLCODE` during validation has `value=0`, except for the transfer from the wallet to the entry point.
+6. No `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` results in an out-of-gas revert.
+7. No referenced address (`CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL`, `EXTCODESIZE`, `EXTCODEHASH` to addresses with `EXTCODESIZE=0`.
+8. Any `GAS` opcode is followed immediately by one of { `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` }.
+9. `EXTCODEHASH` of every address accessed (by any opcode) does not change between first and second simulations of the op.
+10. If `op.initcode.length != 0` , allow only one `CREATE2` opcode call (in the validateUserOp block), otherwise forbid `CREATE2`.
+
+#### Storage access by slots
+
+A contract `A` may reference storage of
+1. Slots of contract `A` address itself. 
+2. Slot `A` on any other address. 
+3. Slots of type `keccak256(A || X)` on any other address. (to cover `mapping(address => value)`, which is usually used for balance in ERC20 tokens). 
+4. Slots of type `keccak256(X || OWN)` on any other address, where OWN is some slot of the previous (third) type 
```

Comment:
30 is too low, considering the modules use case we discussed with Gnosis Safe yesterday.  I'm now unsure whether we should limit it to a fairly high number or not limit it at all (and let the verification gas limit take care of it, which could amount to a few tens of kilobytes in memory).

The use case we discussed is a module (a separate contract) called during validation, which is using its own storage (doesn't delegatecall to the wallet).  Such module can't work as-is, but could be "ported" to 4337 by moving all of its storage to a mapping, in one of two ways:
1. Convert every variable to a mapping.  So `uint256 bla` becomes `mapping(address=>uint256) bla` and is accessed as `bla[msg.sender]` instead of `bla`.
2. Create a struct `Vars` with all the variables and have a `mapping(address=>Vars) vars`.  So `uint256 bla` is now accessed as `vars[msg.sender].bla`.

It enables converting any contract to something that can be used safely during validation, as long as it doesn't use banned opcodes and stays within the verification gas limit.

To support this method, the limit should be high or non-existent.

Am I right that the maximum memory that can be wasted with a reasonable verification gas limit capped at just tens of KB?